Fixed LLM comparator and added pytests for it

ayushi1208 opened the pull request:
*Issue #, if available:*n/a

*Description of changes:*
Updated LLM comparator to use Bedrock call with the ability to customize or use default prompt. Also added test cases.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

ayushi1208 committed:
Fixed llm comparator and added pytests for it

ayushi1208 committed:
Updating dependencies for the LLM comparator

ayushi1208 committed:
Removed unecessary imports in the llm comparator file

ayushi1208 committed:
Updated .gitignore to have .DS_Store in all folders

ayushi1208 committed:
Added name to pyproject.toml

vawsgit commented on pyproject.toml:
```diff
@@ -15,7 +16,8 @@ dependencies = [
     "scipy>=1.10.0",
     "psutil>=5.8.0",
-    "pandas>=1.5.0"
+    "pandas>=1.5.0",
```
For all dependencies I would recommend to add an upper bound limit for the library version. This will prevent the inevitable API breaking change update.

vawsgit reviewed (changes requested):
Probably fix the RUFF/lint errors.

sromoam commented on src/stickler/comparators/llm.py:
```diff
@@ -1,75 +1,106 @@
         self.model_id = model_id
-        self.system_prompt = "You are an evaluation assistant. Carefully decide if the two values are same or not. Respond only with 'TRUE' or 'FALSE', nothing else."
-
-    def compare(self, str1: Any, str2: Any) -> float:
```
This method needs to raise an exception if the user hasn't authenticated to AWS. 
This PR also needs to include instructions for generating credentials or accessing credentials or otherwise authenticating for inference. 

I'm also wondering if we shouldn't use strands agnets here instead of a bedrock invoke, given that strands would let the user configure the model from many providers?

sromoam reviewed (changes requested):
Let's strategize on this PR about a few things
* how we'll provide instructions for the user to get access to a model 
* How we'll test this if we don't have credentials in the github CI pipeline
* If there's a way to make this generic to use any model provider.

vawsgit commented:
> Let's strategize on this PR about a few things
> 
> * how we'll provide instructions for the user to get access to a model
> * How we'll test this if we don't have credentials in the github CI pipeline
> * If there's a way to make this generic to use any model provider.

- For testing this, we might first consider a unit test that mocks the response from the LLM instead of calling the real LLM and consuming resources.
- To make the call compatible with multiple providers, I think using strands is a great idea, but consider this approach:
    - Strands is added as an optional "extra" dependency
    - If it's not installed, then it uses a minimal Bedrock only client implementation.
    - Or...require the users to install the `stickler-eval[llm]` extra if they want to do LLM evaluation, and that will install Strands.

ayushi1208 committed:
Merge branch 'dev' into fix/llm-comparator

ayushi1208 committed:
Merge branch 'dev' into fix/llm-comparator

ayushi1208 committed:
Updated the llm comparator to use strands and updated test cases, pyproject, and included an example scipt

ayushi1208 committed:
Cleaned according to linting rules

ayushi1208 commented:
Please review again @vawsgit @sromoam @adiadd 
I addressed comments.
This now uses strands so that there is option for many model providers. It also mocks the test cases. And it has an add-on of optional dependencies with [llm] in pyproject.toml.

We have some work to do with integrating this comparator in the model from json conversion. Let me know thoughts on whether to make that a part of this PR or the next one.

ayushi1208 committed:
Linting rule removed unused variable

sromoam reviewed (approved):
I'm good with these changes. Would prefer if you can have somebody test hands on, but for a dev commit, also these look fine.

ayushi1208 committed:
Merge branch 'dev' into fix/llm-comparator

ayushi1208 committed:
Fixed the test cases, and verified other tests
